base/precompilation.jl

```diff
@@ -564,7 +591,26 @@ function _precompilepkgs(pkgs::Vector{String},
         end
     end
     if !isempty(circular_deps)
-        @warn """Circular dependency detected. Precompilation will be skipped for:\n  $(join(string.(circular_deps), "\n  "))"""
+        cycles_names = ""
+        for cycle in cycles
+            cycle_str = ""
+            for (i, pkg) in enumerate(cycle)
+                if i == 1
+                    line = "  "
+                elseif i < length(cycle)
+                    line = "  │" * " " ^(i - 2)
+                else
+                    line = "  └" * "─" ^(max(0, i - 3)) * " "
+                end
+                line = color_string(line, :light_black) * full_name(exts, pkg) * "\n"
+                cycle_str *= line
+            end
+            cycles_names *= cycle_str
+        end
```

Comment:
Making a note here, so this isn't forgotten (mostly by me):

This also needs to display any packages in `circular_deps` that weren't in any cycle (i.e. they depend on a dependency cycle, so they were excluded from pre-compile)

Reply:
I just updated it to show these. Example up top